src/test/basics/PerfLog_test.cpp

```diff
@@ -0,0 +1,1075 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2018 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <BeastConfig.h>
+
+#include <ripple/basics/PerfLog.h>
+#include <ripple/beast/unit_test.h>
+#include <ripple/json/json_reader.h>
+#include <ripple/protocol/JsonFields.h>
+#include <ripple/rpc/impl/Handler.h>
+#include <test/jtx/Env.h>
+#include <atomic>
+#include <chrono>
+#include <random>
+#include <string>
+#include <thread>
+
+//------------------------------------------------------------------------------
+
+namespace ripple {
+
+class PerfLog_test : public beast::unit_test::suite
+{
+    enum class WithFile : bool
+    {
+        no = false,
+        yes = true
+    };
+
+private:
+    using path = boost::filesystem::path;
+
+    // We're only using Env for its Journal.  That Journal gives better
+    // coverage in unit tests.
+    test::jtx::Env env_ {*this};
+    beast::Journal j_ {env_.app().journal ("PerfLog_test")};
+
+    // Use this to make calls to random_shuffle() less predictable.
+    std::random_device rd_;
+    std::default_random_engine shuffler_ {rd_()};
+
+    // A PerfLog needs a Parent that is a Stoppable and a function to
+    // call if it wants to shutdown the system.  This class provides both.
+    struct PerfLogParent : public RootStoppable
```

Comment:
I dislike non-trivial inner classes. It's too easy to lose context as to what class I'm in and harder to quickly see interfaces. I'd prefer a detail namespace with the inner class.